src/org/exist/storage/txn/TransactionManager.java

```diff
@@ -85,63 +134,164 @@ public TransactionManager(final BrokerPool pool, final Optional<JournalManager>
+                }
 
-            if(journalManager.isPresent()) {
-                try {
-                    journalManager.get().journal(new TxnStart(txnId));
-                } catch(final JournalException e) {
-                    LOG.error("Failed to create transaction. Error writing to log file.", e);
+                // must NOT begin transaction when another thread is processing system tasks!
+                if (localState == STATE_SYSTEM) {
+                    final long thisThreadId = Thread.currentThread().getId();
+                    if (systemThreadId.compareAndSet(thisThreadId, thisThreadId)) {
+                        // our thread is executing system tasks, allow reentrancy from our thread!
+                        // done... return from CAS loop!
+                        return doBeginTransaction();
+
+                    } else {
+                        // spin whilst another thread executes the system tasks
+                        // sleep a small time to save CPU
+                        Thread.sleep(10);
+                        continue;
+                    }
                 }
+
+                // if we are operational and are not preempted by another thread, begin transaction
+                if (localState >= STATE_IDLE && state.compareAndSet(localState, localState + 1)) {
+                    // done... return from CAS loop!
+                    return doBeginTransaction();
+                }
+            }
+        } catch (final InterruptedException e) {
+            // thrown by Thread.sleep
+            Thread.currentThread().interrupt();
+            //TODO(AR) API should be revised in future so that this is a TransactionException
+            throw new RuntimeException(e);
+        }
+    }
+
+    private Txn doBeginTransaction() {
+        final long txnId = nextTxnId.getAndIncrement();
+        if (journalManager.isPresent()) {
+            try {
+                journalManager.get().journal(new TxnStart(txnId));
+            } catch (final JournalException e) {
+                LOG.error("Failed to create transaction. Error writing to Journal", e);
             }
+        }
 
-            final Txn txn = new Txn(TransactionManager.this, txnId);
-            transactions.put(txn.getId(), new TxnCounter());
-            return txn;
-        });
+        /*
+         * NOTE: we call intentionally increment the txn counter
+         *     to set the counter to 1 to represent the TxnStart
+         *     as that will not be done
+         *     by {@link JournalManager#journal(Loggable)} or
+         *     {@link Journal#writeToLog(loggable)}.
+         */
+        transactions.put(txnId, new TxnCounter().increment());
+
+        return new Txn(this, txnId);
     }
     
     /**
      * Commit a transaction.
      * 
-     * @param txn
-     * @throws TransactionException
+     * @param txn the transaction to commit.
+     *
+     * @throws TransactionException if the transaction could not be committed.
```

Comment:
should `@thows NullPointerException` be added as well?

Reply:
I don't think that is necessary. Almost every Java method is capable of throwing an NPE.